src/NuGet.Clients/NuGet.CommandLine/Common/ProtocolDiagnosticsJsonLogger.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.IO;
+using Newtonsoft.Json;
+using NuGet.Protocol.Utility;
+
+namespace NuGet.CommandLine
+{
+    internal sealed class ProtocolDiagnosticsJsonLogger : IDisposable
+    {
+        private object _lock;
+        private StreamWriter _streamWriter;
+        private JsonWriter _jsonWriter;
+
+        internal ProtocolDiagnosticsJsonLogger(string fileName)
+        {
+            _lock = new object();
+            _streamWriter = new StreamWriter(fileName);
+            _jsonWriter = new JsonTextWriter(_streamWriter);
+            _jsonWriter.Formatting = Formatting.Indented;
+            _jsonWriter.WriteStartArray();
+        }
+
+        internal void OnEvent(ProtocolDiagnosticEvent pde)
+        {
+            if (_lock == null)
+            {
+                throw new ObjectDisposedException(nameof(ProtocolDiagnosticsJsonLogger));
+            }
+
+            var timestamp = pde.Timestamp.ToString("o", CultureInfo.InvariantCulture);
+
+            lock (_lock)
+            {
+                _jsonWriter.WriteStartObject();
+
+                _jsonWriter.WritePropertyName("timestamp");
+                _jsonWriter.WriteValue(timestamp);
+
+                _jsonWriter.WritePropertyName("source");
+                _jsonWriter.WriteValue(pde.Source);
+
+                _jsonWriter.WritePropertyName("url");
```

Comment:
Might be interesting to include HTTP method and reason phrase too since reason phrase is NuGet's error message field.